pageserver/src/tenant/timeline.rs

```diff
@@ -1221,70 +1232,161 @@ impl Timeline {
+        match *guard {
+            Running | Success => return,
+            NotStarted | FailedWillRetryNextTime => *guard = Running,
+        }
+        drop(guard);
+        // We need to start the computation task.
+        let self_clone = Arc::clone(self);
+        task_mgr::spawn(
+            task_mgr::BACKGROUND_RUNTIME.handle(),
+            task_mgr::TaskKind::InitialLogicalSizeCalculation,
+            Some(self.tenant_id),
+            Some(self.timeline_id),
+            "initial size calculation",
+            false,
+            // NB: don't log errors here, task_mgr will do that.
+            async move {
+                let res = self_clone.logical_size_calculation_task(init_lsn).await;
+                let new_state = match res {
+                    Ok(_) => Success,
+                    Err(_) => FailedWillRetryNextTime,
+                };
+                let mut state = self_clone.initial_size_computation_state.lock().unwrap();
+                if *state != Running {
+                    // Should be unreachable, but no reason to crash the pageserver. Don't touch anything.
+                    error!("expecting initial size computation task to be in state {Running:?}, got {state:?}")
+                } else {
+                    *state = new_state;
+                }
+
+                let calculated_size = res.context("Failed to calculate logical size")?;
+                match self_clone
+                    .current_logical_size
+                    .initial_logical_size
+                    .set(calculated_size)
+                {
+                    Ok(()) => (),
+                    Err(existing_size) => {
+                        // This shouldn't happen because we use self.initial_size_computation_running to ensure exlusivity here.
+                        // But if it happens, just complain & report success so there are no further retries.
+                        error!("Tried to update initial timeline size value to {calculated_size}, but the size was already set to {existing_size}, not changing")
+                    }
+                }
+                Ok(())
+            },
+        );
+    }
+
+    pub fn spawn_ondemand_logical_size_calculation(
+        self: &Arc<Self>,
+        init_lsn: Lsn,
+    ) -> oneshot::Receiver<anyhow::Result<u64>> {
+        let (sender, receiver) = oneshot::channel();
+        let self_clone = Arc::clone(self);
+        task_mgr::spawn(
+            task_mgr::BACKGROUND_RUNTIME.handle(),
+            task_mgr::TaskKind::InitialLogicalSizeCalculation,
+            Some(self.tenant_id),
+            Some(self.timeline_id),
+            "ondemand logical size calculation",
+            false,
+            async move {
+                let res = self_clone.logical_size_calculation_task(init_lsn).await;
+                let _ = sender.send(res).ok();
+                Ok(()) // Receiver is responsible for handling errors
+            },
+        );
+        receiver
```

Comment:
Is it really worth it to use task_mgr here?

Note we also don't cancel this task if the oneshot receiver goes away, something which would not be an issue with cancellation token's dropguard and just returning the value through JoinHandle.

Reply:
We definitely need to use task_mgr so that tenant_detach / timeline_delete don't delete the on-disk state while the calculation is still ongoing.
They rely on task_mgr::shutdown and this task is timeline scoped.
It's the right tool for the job here.

> Note we also don't cancel this task if the oneshot receiver goes away, something which would not be an issue with cancellation token's dropguard and just returning the value through JoinHandle.

If your https://github.com/neondatabase/neon/pull/3124 could make the cancellation token available to the task code, then we could funnel that through here.

Reply:
#3124 has now been merged, did you update this afterwards, are you still planning? Please ping me if you do on slack.

> They rely on task_mgr::shutdown and this task is timeline scoped.
> It's the right tool for the job here.

I dislike it because the joinhandles are in general so useful tool this is just so foreign. But agreed cannot really see a way around that in this model. Also I failed to see the chance of detach "race" here earlier.

Reply:
I'll merge this PR without exposing the cancellation token - we're using `task_mgr::shutdown_watcher` and I think that's the way to go even after #3124 . We can replace `shutdown_watcher` with a getter for the cancellation token in the future.